spring-cloud-dataflow-shell-core/src/main/java/org/springframework/cloud/dataflow/shell/config/DataFlowShell.java

```diff
@@ -16,25 +16,125 @@
 
 package org.springframework.cloud.dataflow.shell.config;
 
+import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.cloud.dataflow.rest.client.DataFlowOperations;
+import org.springframework.cloud.dataflow.shell.Target;
+import org.springframework.cloud.dataflow.shell.TargetCredentials;
+import org.springframework.cloud.dataflow.shell.TargetHolder;
+import org.springframework.cloud.dataflow.shell.command.support.OpsType;
+import org.springframework.cloud.dataflow.shell.command.support.RoleType;
 import org.springframework.stereotype.Component;
 
 /**
  * REST client component that holds all the available operations for
  * communicating with the Spring Cloud Data Flow REST server.
  *
  * @author Ilayaperumal Gopinathan
+ * @author Gunnar Hillert
  */
 @Component
 public class DataFlowShell {
 
 	private DataFlowOperations dataFlowOperations;
 
+	private TargetHolder targetHolder;
+
 	public DataFlowOperations getDataFlowOperations() {
 		return dataFlowOperations;
 	}
 
+	@Autowired
+	public void setTargetHolder(TargetHolder targetHolder) {
+		this.targetHolder = targetHolder;
+	}
+
 	public void setDataFlowOperations(DataFlowOperations dataFlowOperations) {
 		this.dataFlowOperations = dataFlowOperations;
 	}
+
+	/**
+	 * This method performs various access checks and only if {@code true} is returned, the shell should show the associated
+	 * commands. 2 checks are performed:
+	 *
+	 * <ul>
+	 *   <li>Does the desired operation indicated by the {@link OpsType} exist?
+	 *   <li>If the operation exist, does the user have the necessary credentials?
+	 * </ul>
+	 *
+	 * <b>What are valid credentials?</b>
+	 *
+	 * The way the passed-in {@link RoleType}s are treated depends on the servers security settings:
+	 *
+	 *<ul>
+	 *   <li>{@link Target#isAuthenticationEnabled()} Is authentication enabled?
+	 *   <li>{@link Target#isAuthorizationEnabled()} If authentication is enabled, is authorization enabled?
+	 *   <li>{@link Target#isAuthenticated()} Is the user authenticated?
+	 *   <li>Only if {@link Target#isAuthorizationEnabled()} and {@link Target#isAuthenticated()} is {@code true}
+	 *   will the {@link RoleType} checked against {@link TargetCredentials#getRoles()}.
+	 * </ul>
+	 *
+	 * @param role Can be null.
+	 * @param opsType Must not be null.
+	 * @return If true the indicated operation is accessible and the user has all credentials to execute the operation.
+	 */
+	public boolean hasAccess(RoleType role, OpsType opsType) {
+		if (this.dataFlowOperations != null) {
+			boolean operationsAvailable = false;
+			switch(opsType) {
+				case AGGREGATE_COUNTER:
+					operationsAvailable = this.dataFlowOperations.aggregateCounterOperations() != null;
```

Comment:
I feel like all these "available" checks belong to the commands, where they were initially. Why not `&&` them in the `@CliAvailabilityIndicator` methods?